client/allocrunner/taskrunner/volume_hook.go

```diff
@@ -0,0 +1,123 @@
+package taskrunner
+
+import (
+	"context"
+	"fmt"
+
+	log "github.com/hashicorp/go-hclog"
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/hashicorp/nomad/plugins/drivers"
+)
+
+type volumeHook struct {
+	alloc  *structs.Allocation
+	runner *TaskRunner
+	logger log.Logger
+}
+
+func newVolumeHook(runner *TaskRunner, logger log.Logger) *volumeHook {
+	h := &volumeHook{
+		alloc:  runner.Alloc(),
+		runner: runner,
+	}
+	h.logger = logger.Named(h.Name())
+	return h
+}
+
+func (*volumeHook) Name() string {
+	return "volumes"
+}
+
+func validateHostVolumes(requested map[string]*structs.VolumeRequest, client map[string]*structs.ClientHostVolumeConfig) error {
+	var result error
+
+	for n, req := range requested {
+		if req.Volume.Type != "host" {
+			continue
+		}
+
+		cfg, err := structs.ParseHostVolumeConfig(req.Config)
+		if err != nil {
+			result = multierror.Append(result, fmt.Errorf("failed to parse config for %s: %v", n, err))
+		}
+
+		_, ok := client[cfg.Source]
+		if !ok {
+			result = multierror.Append(result, fmt.Errorf("missing %s", cfg.Source))
+		}
+	}
+
+	return result
+}
+
+func (h *volumeHook) hostVolumeMountConfigurations(vmounts []*structs.VolumeMount, volumes map[string]*structs.VolumeRequest, client map[string]*structs.ClientHostVolumeConfig) ([]*drivers.MountConfig, error) {
```

Comment:
I'm getting lost tracking the the distinction between volume mount, volume config, mount config, and volume mount config :).  I'd suggest adding a comment somewhere to explain what each type is meant to represent.

Also, I'd love to clarify the parameter names here, `client`.  Maybe naming the parameter based on source of value would be useful if they are all volumes of some kind.

Reply:
ok - So I have a better handle now - Let me phrase them as best as I understand and have me confirm:

* `volume` is the filesystem thingy that is unit of scheduling to be assigned to an alloc
* `host volume` is the client configuration of a volume.  `HostVolume` fields and variables in this PR refer to client config values of type `ClientHostVolumeConfig` (or derived from it) and never to the volumes instances themselves.
* `volume mount` refers to the task level configuration for how the volume is mounted to the task filesystem.
* `mount config` refers to the concrete OS bind specification that executors will use to actually bind mount?

Is this roughly correct?  Would be nice to call that out in the struct documentation?